#40 - Connecting GitHub API for Researchers to search for new participants

dylHall opened the pull request:
The preliminary connection of the GitHub API to obtain new Participants is ready to be reviewed/pulled in - because other issues depend on this change, I’d like to begin reviewing these changes sooner rather than later. While we perform a code review and get feedback on this, I’ll investigate the feasibility of adding tests for these changes (although these might be more suited to a separate PR). 

**Please refer to some of the notes I made on #40 for some background on the workflow for using this feature.** 

### Summary of Front End Changes
Added a Form to retrieve information from the Researcher to send a GitHub API request. This includes introducing a specific model to back the form, and some new services for sending an internal REST query to our backend server. All additional files have been placed within the \app\entities\participant\githubAPI folder, but some changes to the files in the Participants entity folder were required to introduce routing to the new modal dialogue displaying the form.

### Some Specifics of Front End Changes
- The GitHub user search form itself is controlled by the `GitHubQueryDialogComponent `object, which is responsible for sending the request to the java resource handler. It subscribes to the response to be able to reload the Participants that have been added, but some additional work could be done in this component to convey better processing information to the user. The component and template can be found in the participant-github-query-dialog.component.[ts/html]
- The `GitHubQueryPopupComponent`/`GitHubPopupService `are used to manage the process of instantiating the model underlying the form, and producing the “popup” nature of the form.
- GitHubAPIRequest object represents the backing model of the form, which translates to having a member variable for each of the form fields that are displayed. These are modified by using the ngModel attribute in the html template.
- GitHubBackendCall provides the service of copying the `GitHubAPIRequest `object, and producing the call to the java resource handler class (with the `GitHubAPIRequest `object as the body of the message)
- Changes made to the routing/module of the Participants entity folder (to route to the new form) and added a button on the page to open up the modal - thus had a few changes to the html document for the Participants page as well.

### Summary of Back End Changes
Because of the CORS issues I encountered making an external API request from the front end, I opted to make the GitHub API call from the backend to promote interoperability. This change includes introducing a new Resource Handler, and some extra objects for marshaling and controlling logic flow. I’ve also added a new Maven dependency to a GitHub API java package (`kohsuke.github`) to allow for an abstracted view of the API and provide a domain model in a sense.

### Some Specifics of Back End Changes
- The resource handler is the `GitHubResource `class, which has been annotated to receive internal requests through the resource URL - `api/gitHubQuery.` It currently only has one HTTP handler method for a POST, taking a `GitHubAPIRequest `from the body of the HTML message. It will perform the search with the parameters entered into the form, and will the construct Participants and add them to the database to be displayed. Please refer to the comments/annotations in this class for additional information about I/O and Todos.
- Note: as an interim solution, **a fake email containing the users' GitHub username has been added to the constructed participant** (instead of the actual obtained email from the search query), the occupation has been set to the “company” field of a user (no actual occupation information available), and the programming language set is the first primary language found for the users' repositories. **Retrieving contribution counts have been disabled by default (checkbox form), as this is extremely network intensive and will potentially max out an API call allocation quickly (limitations of REST API’s – could investigate GraphQL to optimize this)**. API calls limited to 60/hour currently, but can be increased to 5000 if login with an authentication key.
- `GitHubAPIRequest`/`GitHubUserType `are just backend mappings for information sent from the front end. These have automatic marshaling despite not having any annotations.

Although I’ve added some preliminary precautions to failures, this is quite a large change so I’m sure others will find bugs – let me know if you find some or have any general queries!

dylHall commented:
@crat019 @KijongHan this is ready for code review

crat019 commented on src/main/java/org/project36/qualopt/service/github/GitHubAPIRequest.java:
```diff
@@ -0,0 +1,110 @@
+
+    private int muleId;
+
+    private GitHubUserType userType; // convert to Enum later
```
Is this comment redundant now? It looks like you have changed it to an enum

dylHall replied:
Ah nice find - will remove that comment now!

KijongHan commented on src/main/java/org/project36/qualopt/service/github/GitHubAPIRequest.java:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    @Override
+    public String toString() {
```
Might be a good idea to include all the fields in this toString method. Missing muleid and keyword

dylHall replied:
Yup true, will add them in

dylHall commented:
@crat019 @KijongHan addressed comments and added last few touch-ups - so could you have another look at this PR and comment/approve?

@kblincoe will add a testing class after the changes in #108 have been merged, as these changes seem to be working successfully with the spring boot upgrade right now (compiled and was able to execute queries/make participants)

KijongHan commented:
Looks good Dylan.

crat019 commented:
Looks Good To Me 👍

dylHall commented:
@softeng-701 Went through code with my team, tested on separate computers and fixed the last few bugs I could see - ready to be merged now.
